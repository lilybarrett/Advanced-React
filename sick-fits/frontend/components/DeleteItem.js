import React, { Component } from 'react'
import { Mutation } from "react-apollo";
import { ALL_ITEMS_QUERY } from "./Items";
import gql from "graphql-tag";

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id 
        }
    }
`;

class DeleteItem extends Component {
    update = (cache, payload) => {
        // steps to manually update the client cache so it matches the data that's actually on the server
        // read the cache for the items we want -- in this case, we want to make a query for "all the items" from the cache
        const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
        console.log(data);
        // filter the deleted item out of the page
        data.items = data.items.filter((item) => item.id !== payload.data.deleteItem.id);
        // put the items back!
        cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
    }
    render() {
        return (
            <Mutation
                mutation={DELETE_ITEM_MUTATION}
                variables={{ id: this.props.id }}
                update={this.update}>
                {( deleteItem, { error } ) => (
                    <button onClick={() => {
                        if (confirm("Are you sure you want to delete this item?")) {
                            deleteItem();
                        }
                    }}>
                        {this.props.children}
                    </button>
                )}
            </Mutation>
        )
    }
}

export default DeleteItem;

