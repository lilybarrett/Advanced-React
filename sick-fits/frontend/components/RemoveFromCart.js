import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { CURRENT_USER_QUERY } from './User';

const BigButton = styled.button`
    font-size: 3rem;
    background: none;
    border: 0;
    &:hover {
        color: ${props => props.theme.red};
    }
    cursor: pointer;
`;

const REMOVE_FROM_CART_MUTATION = gql`
    mutation removeFromCart($id: ID!) {
        removeFromCart(id: $id) {
            id 
        }
    }
`;

class RemoveFromCart extends React.Component {
    static propTypes = {
        id: PropTypes.string.isRequired
    }
    // this gets called as soon as we get a response back from the server after performing the mutation
    update = (cache, payload) => {
        // cache is our Apollo cache 
        // payload is actual dump of information that's returned from the server
        // first, read the cache
        const data = cache.readQuery({
            query: CURRENT_USER_QUERY
            // this allows us to get the current user's cart
        });
        // remove the item from the cart
        const cartItemId = payload.data.removeFromCart.id;
        // for this reason, we purposefully made sure our removeFromCart mutation would return the ID for an item
        data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);
        // write it back to the cache
        cache.writeQuery({ query: CURRENT_USER_QUERY, data })
    }

    render () {
        const { id } = this.props;
        return (
            <Mutation
                mutation={REMOVE_FROM_CART_MUTATION}
                variables={{ id }}
                update={this.update}
                optimisticResponse={{
                    // this allows us to immediately run the update fn rather than waiting for the response from the server
                    __typename: 'Mutation',
                    removeFromCart: {
                        __typename: 'CartItem',
                        id: this.props.id,
                    }
                }}>
                {(removeFromCart, { loading, error }) => (
                    <BigButton
                        disabled={loading}
                        onClick={() => {
                            removeFromCart().catch(err => alert(err.message))
                        }}
                        title="Delete Item">
                        &times;
                    </BigButton>
                )}
            </Mutation>
        )
    }
}

export default RemoveFromCart;

