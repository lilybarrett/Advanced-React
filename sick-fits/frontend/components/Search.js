import React from 'react';
import Downshift from 'downshift';
import Router from 'next/router';
import { ApolloConsumer } from 'react-apollo';