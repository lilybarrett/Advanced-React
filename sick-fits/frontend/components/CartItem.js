import React from 'react';
import formatMoney from '../lib/formatMoney';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import RemoveFromCart from './RemoveFromCart';

const CartItemStyles = styled.li`
    padding: 1rem 0;
    border-bottom: 1px solid ${props => props.theme.lightgrey};
    display: grid;
    align-items: center;
    grid-template-columns: auto 1fr auto;
    img {
        margin-right: 10px;
    }
    h3, p {
        margin: 0;
    }
`;

const CartItem = ({ cartItem }) => {
    if (!cartItem.item) return (
        <CartItemStyles>
            This item has been removed from the shop<RemoveFromCart id={cartItem.id}/>
        </CartItemStyles>
    )
    return (
        <CartItemStyles>
            <img width='100' src={cartItem.item.image} alt={cartItem.item.title} />
            <div className='cart-item-details'>
                <h3>{cartItem.item.title}</h3>
            </div>
            <p>
                {formatMoney(cartItem.item.price * cartItem.quantity)}
                {' - '}
                <em>
                    {cartItem.quantity} &times; {formatMoney(cartItem.item.price)}
                </em>
            </p>
            <RemoveFromCart id={cartItem.id} />
        </CartItemStyles>
    )
}

CartItem.propTypes = {
    cartItem: PropTypes.shape({
        id: PropTypes.string,
        quantity: PropTypes.number,
        item: PropTypes.shape({
            title: PropTypes.string,
            price: PropTypes.number,
            image: PropTypes.string
        })
    })
}

export default CartItem;