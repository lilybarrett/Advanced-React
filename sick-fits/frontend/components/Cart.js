import React from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag'; 
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import User from './User';
import CartItem from './CartItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

export const LOCAL_STATE_QUERY = gql`
    query {
        cartOpen @client 
        # the @client directive tells Apollo not to grab this from the API, but from our local data store in Apollo Client

    }
`;

export const TOGGLE_CART_MUTATION = gql`
    mutation {
        toggleCart @client
    }
`;

const Cart = () => {
    return (
        <User>
            {({ data: { me }}) => {
                console.log(me);
                if (!me) return null;
                return (
                    <Mutation mutation={TOGGLE_CART_MUTATION}>
                        {( toggleCart ) => (
                            <Query query={LOCAL_STATE_QUERY}>
                                {({ data }) => (
                                    <CartStyles open={data.cartOpen}>
                                        <header>
                                            <CloseButton onClick={toggleCart} title="close">
                                                &times;
                                            </CloseButton>
                                            <Supreme>
                                                {me.name}'s Cart
                                            </Supreme>
                                            <p>
                                                You have {me.cart.length} item{me.cart.length > 1 && `s`} in your cart
                                            </p>
                                        </header>
                                        <ul>
                                           {me.cart.map((cartItem) => <CartItem key={cartItem.id} cartItem={cartItem} />)} 
                                        </ul>
                                        <footer>
                                            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
                                            <SickButton>Checkout</SickButton>
                                        </footer>
                                    </CartStyles>
                                )}
                            </Query>
                        )}
                    </Mutation>
                )
            }}
        </User>
    )
}

export default Cart;

