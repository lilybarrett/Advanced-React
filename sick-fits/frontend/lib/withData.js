import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    // local data store that we're passing around using Apollo -- as opposed to Redux or Context
    clientState: {
      resolvers: {
        Mutation: {
          // underscore -- Wes isn't sure what this represents
          // variables (we don't actually have any in this case)
          // cache -- destructured out of Apollo Client -- where we will store cartOpen value
          toggleCart(_, variables, { cache }) {
            // read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            })
            // write the cart state to the opposite toggle state
            const data = {
              data: { cartOpen: !cartOpen }
            }
            cache.writeData(data);
            console.log(cartOpen);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false,
      },
    }
  });
}

export default withApollo(createClient);
