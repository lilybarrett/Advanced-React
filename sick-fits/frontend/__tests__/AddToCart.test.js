import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import AddToCart, { ADD_TO_CART_MUTATION } from '../components/AddToCart';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import { ApolloConsumer } from 'react-apollo';

const mocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: {
            me: {
                ...fakeUser(),
                cart: [],
            }
        }}
    },
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: {
            me: {
                ...fakeUser(),
                cart: [fakeCartItem()],
            }
        }}
    },
    {
        request: { query: ADD_TO_CART_MUTATION, variables: { id: 'abc123' }},
        result: {
            data: {
                addToCart: {
                    ...fakeCartItem(),
                    quantity: 1,
                }
            }
        }
    }
]

describe('<AddToCart />', () => {
    it('renders and matches the snapshot', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <AddToCart id='abc123'/>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const button = wrapper.find('button');
        expect(toJSON(button)).toMatchSnapshot();
    });

    it('adds an item to cart when clicked', async () => {
        let apolloClient;
        const wrapper = mount(
           <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                    {(client) => {
                        apolloClient = client;
                        return <AddToCart id='abc123' />
                    }}
                </ApolloConsumer>
           </MockedProvider>
       );
       await wait();
       wrapper.update();
       const { data: { me }} = await apolloClient.query({ query: CURRENT_USER_QUERY });
    //    console.log(me);
       expect(me.cart).toHaveLength(0);
       // add an item to the cart
       const button = wrapper.find('button[data-test="add-button"]');
       button.simulate('click');
       await wait();
       // check that item is in cart
       const { data: { me: me2 }} = await apolloClient.query({ query: CURRENT_USER_QUERY });
       console.log(me2);
    //    expect(me2.cart).toHaveLength(1);
    });
});