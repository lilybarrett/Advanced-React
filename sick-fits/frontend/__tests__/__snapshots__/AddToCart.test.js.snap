// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart /> renders and matches the snapshot 1`] = `
<button
  data-test="add-button"
  disabled={false}
  onClick={[Function]}
>
  Add
   To Cart
</button>
`;
